slack_sdk/socket_mode/async_client.py

```diff
@@ -58,15 +60,24 @@ async def issue_new_wss_url(self) -> str:
                 self.logger.error(f"Failed to retrieve WSS URL: {e}")
                 raise e
 
+    async def is_connected(self) -> bool:
+        return False
+
     async def connect(self):
         raise NotImplementedError()
 
     async def disconnect(self):
         raise NotImplementedError()
 
-    async def connect_to_new_endpoint(self):
-        self.wss_uri = await self.issue_new_wss_url()
-        await self.connect()
+    async def connect_to_new_endpoint(self, force: bool = False):
+        try:
+            await self.connect_operation_lock.acquire()
```

Comment:
The lock acquisition and release inside this method is the main change in this PR. With this way, other concurrent method calls are synchronized. The client can safely check `sef.is_connected()` inside and decide if the client still needs to replace the WSS URL and the underlying WebSocket session.